pkg/internalapis/featurestates/featurestates.go

```diff
@@ -0,0 +1,418 @@
+		return
+	}
+	if fssConfigMap.Name == supervisorFeatureStatConfigMapName &&
+		fssConfigMap.Namespace == supervisorFeatureStateConfigMapNamespace {
+		log.Infof("observed fss add: fss: %+v", fssConfigMap.Data)
+		var err error
+		var featureStates []featurestatesv1alpha1.FeatureState
+		for feature, state := range fssConfigMap.Data {
+			var featureState featurestatesv1alpha1.FeatureState
+			featureState.Name = feature
+			featureState.Enabled, err = strconv.ParseBool(state)
+			if err != nil {
+				log.Warnf("failed to parse feature state value: %v for feature: %q", state, feature)
+			}
+			featureStates = append(featureStates, featureState)
+		}
+		pendingCRUpdatesObj.enqueueFeatureStateUpdatesForAllWorkloadNamespaces(ctx, featureStates)
+	}
+}
+
+// configMapUpdated is called when configmap is updated
+func configMapUpdated(oldObj, newObj interface{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ctx = logger.NewContextWithLogger(ctx)
+	log := logger.GetLogger(ctx)
+
+	newfssConfigMap, ok := newObj.(*v1.ConfigMap)
+	if newfssConfigMap == nil || !ok {
+		log.Warnf("configMapUpdated: unrecognized old object %+v", newObj)
+		return
+	}
+	oldfssConfigMap, ok := oldObj.(*v1.ConfigMap)
+	if oldfssConfigMap == nil || !ok {
+		log.Warnf("configMapUpdated: unrecognized new object %+v", newObj)
+		return
+	}
+
+	if newfssConfigMap.Name == supervisorFeatureStatConfigMapName &&
+		newfssConfigMap.Namespace == supervisorFeatureStateConfigMapNamespace {
+		if reflect.DeepEqual(newfssConfigMap.Data, oldfssConfigMap.Data) {
+			return
+		}
+		log.Infof("observed fss update: oldfss: %+v newfss: %+v", oldfssConfigMap.Data, newfssConfigMap.Data)
+		var err error
+		var featureStates []featurestatesv1alpha1.FeatureState
+		for feature, state := range newfssConfigMap.Data {
+			var featureState featurestatesv1alpha1.FeatureState
+			featureState.Name = feature
+			featureState.Enabled, err = strconv.ParseBool(state)
+			if err != nil {
+				log.Warnf("failed to parse feature state value: %v for feature: %q", state, feature)
+			}
+			featureStates = append(featureStates, featureState)
+		}
+		pendingCRUpdatesObj.enqueueFeatureStateUpdatesForAllWorkloadNamespaces(ctx, featureStates)
+	}
+}
+
+// configMapDeleted is called when config-map is deleted
+func configMapDeleted(obj interface{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ctx = logger.NewContextWithLogger(ctx)
+	log := logger.GetLogger(ctx)
+
+	fssConfigMap, ok := obj.(*v1.ConfigMap)
+	if fssConfigMap == nil || !ok {
+		log.Warnf("configMapDeleted: unrecognized object %+v", obj)
+		return
+	}
+	if fssConfigMap.Name == supervisorFeatureStatConfigMapName &&
+		fssConfigMap.Namespace == supervisorFeatureStateConfigMapNamespace {
+		log.Warnf("supervisor feature switch state configmap %q from the namespace: %q is deleted", supervisorFeatureStatConfigMapName, supervisorFeatureStateConfigMapNamespace)
```

Comment:
Dont you have to clear the map since FSS itself is not there?

Reply:
If we clear the map, then after fss config-map is deleted accidentally, we will end up setting `false` in all existing workload namespaces and all new workload namespaces added after deleting fss config-map will have all features set as false. So just kept this as a warning message. 

Ideally, SV admin should never delete this internal config-map.

Reply:
If FSS config map is deleted, there is no need to perform any replication. We can keep the logic simple by just clearing the map.

Reply:
updating configMapDeleted as below to clear in-memory cached feature switch states.

```
// configMapDeleted is called when config-map is deleted
func configMapDeleted(obj interface{}) {
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
	ctx = logger.NewContextWithLogger(ctx)
	log := logger.GetLogger(ctx)

	fssConfigMap, ok := obj.(*v1.ConfigMap)
	if fssConfigMap == nil || !ok {
		log.Warnf("configMapDeleted: unrecognized object %+v", obj)
		return
	}
	if fssConfigMap.Name == supervisorFeatureStatConfigMapName &&
		fssConfigMap.Namespace == supervisorFeatureStateConfigMapNamespace {
		log.Warnf("supervisor feature switch state configmap %q from the namespace: %q is deleted", supervisorFeatureStatConfigMapName, supervisorFeatureStateConfigMapNamespace)
		pendingCRUpdatesObj.lock.Lock()
		defer pendingCRUpdatesObj.lock.Unlock()
		pendingCRUpdatesObj.latestFeatureStates = make([]featurestatesv1alpha1.FeatureState, 0)
		for namespace := range pendingCRUpdatesObj.namespaceUpdateMap {
			pendingCRUpdatesObj.namespaceUpdateMap[namespace] = false
		}
	}
}
```

Reply:
Also updating processPendingCRUpdates() to handle case when empty fss is observed.

```golang
// processPendingCRUpdates helps process pending CR updates at regular interval
func (pendingCRUpdatesObj *pendingCRUpdates) processPendingCRUpdates() {
	ticker := time.NewTicker(crUpdateRetryInterval)
	for range ticker.C {
		func() {
			pendingCRUpdatesObj.lock.Lock()
			defer pendingCRUpdatesObj.lock.Unlock()

			ctx, log := logger.GetNewContextWithLogger()
			if len(pendingCRUpdatesObj.latestFeatureStates) == 0 {
				log.Warnf("empty feature states observed. feature state config-map might be deleted.")
				return
			}
.
.

```